cs_api_io/cs_api_io.ml

```diff
@@ -1,79 +1,53 @@
-  Conduit_lwt_unix.init ~verify () >|= fun ctx ->
-  Cohttp_lwt_unix.Client.custom_ctx ~ctx ()
+module Response = struct
+  type t =
+    { code : int
+    ; response : string ref }
+end
 
-let add_part ~multipart {Api.Part.name; content} =
-  match content with
-  | Direct value ->
-    Multipart_form_writer.add_form_element ~name ~value multipart
-  | File {path; size = _} ->
-    Multipart_form_writer.add_file_from_disk ~name ~path multipart
+let response_accumulator_factory () =
+  let resp = ref "" in
+  let resp_callback s =
+    resp := !resp ^ s;
+    String.length s
+  in
+  (resp, resp_callback)
 
-let handle_raw str = Lwt_result.return ([], Lwt_stream.of_list [str])
+let set_headers curl header =
+  Curl.set_httpheader curl (List.map (fun (h, v) -> h ^ ": " ^ v) header)
 
-let handle_multipart parts =
-  let open Lwt_result.Infix in
-  let multipart =
-    List.fold_left
-      (fun multipart part -> add_part ~multipart part)
-      (Multipart_form_writer.init ())
-      parts
-  in
-  Multipart_form_writer.r_body multipart >>= fun body ->
-  Multipart_form_writer.r_headers multipart >|= fun headers -> (headers, body)
+let set_part {Api.Part.name; content} =
+  match content with
+  | Direct s -> Curl.CURLFORM_CONTENT (name, s, Curl.DEFAULT)
+  | File {path; _} -> Curl.CURLFORM_FILECONTENT ("file", path, Curl.DEFAULT)
+
+let set_multipart curl parts = List.map set_part parts |> Curl.set_httppost curl
 
 let send_request_exn ~verify {Api.Request.url; header; method_; data} =
-  let open Lwt.Infix in
-  let method_str =
+  Curl.global_init Curl.CURLINIT_GLOBALALL;
+  let curl = Curl.init () in
+  let (response, response_callback) = response_accumulator_factory () in
+  Curl.set_url curl url;
+  Curl.set_sslverifypeer curl verify;
+  Curl.set_writefunction curl response_callback;
+  set_headers curl header;
+  let _ =
     match method_ with
-    | Get -> "GET"
-    | Post -> "POST"
+    | Get -> Curl.set_httpget curl true
+    | Post -> (
+      match data with
+      | Raw str -> Curl.set_postfields curl str
+      | Multipart parts -> set_multipart curl parts )
   in
-  Printf.printf "Request: %s %s\n%!" method_str url;
-  let headers = Cohttp.Header.add_list (Cohttp.Header.init ()) header in
-  let url = Uri.of_string url in
-  get_ctx ~verify >>= fun ctx ->
-  let open Lwt_result.Infix in
-  match method_ with
-  | Get -> Lwt_result.ok (Cohttp_lwt_unix.Client.get ~ctx ~headers url)
-  | Post ->
-    ( match data with
-    | Raw str -> handle_raw str
-    | Multipart parts -> handle_multipart parts )
-    >>= fun (extra_headers, body) ->
-    let headers = Cohttp.Header.add_list headers extra_headers in
-    let body = Cohttp_lwt.Body.of_stream body in
-    Lwt_result.ok (Cohttp_lwt_unix.Client.post ~ctx ~body ~headers url)
+  try
+    Curl.perform curl;
+    Lwt_result.return {Response.code = Curl.get_responsecode curl; response}
+  with Curl.CurlException (_, _, s) ->
+    Lwt_result.fail ("HTTP exception : " ^ s)
```

Comment:
- What about "HTTP error"?   I think would make more sense to the user than "exception" in the context of HTTP.
- As you probably know now, unlike in French, there's no space before a colon or semi-colon character in English.

Reply:
I'll never get used to the colons :(... Fixed.